### PR TITLE
`Integrated code lifecycle`: Trim docker image name to avoid white space issues

### PR DESCRIPTION
**Pull request by krusche**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I **strictly** followed the [client coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context
It can happen that the docker image has a leading or trailing whitespace which will lead to build errors that are currently hard to detect for instructors

### Description
We make sure to trim the docker image name in several places to avoid issues with existing database entries and to prevent new database entries with unnecessary whitespace

Additionally, this PR improves a couple of code quality issues:
* Unify the use of ObjectMapper (instead of Gson) to avoid maintaining multiple libraries in the context of programming exercises
* Convert DTOs to records (and add additional TODOs for other cases)
* Use consistent Json annotations for DTOs
* Prefer `Strings.isNullOrEmpty(string)` over `string.isBlank()` to avoid null pointer exceptions

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Create a new programming exercise, customize the build and add a trailing or leading white space into the docker image --> Save
2. The build should still occur. There should be no whitespace in the database (buildConfiguration json) and if you edit again, there should be no white space

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### CODE REVIEW

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts [hunk 1]
```diff
@@ -600,6 +600,11 @@ export class ProgrammingExerciseUpdateComponent implements AfterViewInit, OnDest
      * Saves the programming exercise with the provided input
      */
     saveExercise() {
+        // trim potential whitespaces that can lead to issues
+        if (this.programmingExercise.windFile?.metadata?.docker?.image) {
+            this.programmingExercise.windFile.metadata.docker.image = this.programmingExercise.windFile.metadata.docker.image.trim();
+        }
+        this.programmingExercise.windFile?.metadata?.docker?.image?.trim();
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:607]: `trim()` does not modify the string.
```suggestion
```

File: src/main/webapp/app/exercises/programming/manage/update/update-components/custom-build-plans/programming-exercise-docker-image/programming-exercise-docker-image.component.html [hunk 2]
```diff
@@ -11,6 +11,6 @@
         name="dockerImage"
         id="field_dockerImage"
         [(ngModel)]="dockerImage"
-        (ngModelChange)="dockerImageChange.emit($event)"
+        (ngModelChange)="dockerImage = $event.trim(); dockerImageChange.emit(dockerImage)"
```
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/manage/update/update-components/custom-build-plans/programming-exercise-docker-image/programming-exercise-docker-image.component.html:14]: Can you move this into a method in the template? Imo we should only have method-delegation in html templates.

Comment (krusche) [reply src/main/webapp/app/exercises/programming/manage/update/update-components/custom-build-plans/programming-exercise-docker-image/programming-exercise-docker-image.component.html:14]: This was suggested by ChatGPT :-) I also trim it in the corresponding setter, so we could probably remove it here again

**Review by Strohgelaender — DISMISSED** [review]

**Review by Strohgelaender — DISMISSED** [review]
code lgtm

**Review by pzdr7 — DISMISSED** [review]
Code looks good 👍 
Tested on TS1. The image name was trimmed as expected

Template & Solution build worked fine as well

**Review by Hialus — DISMISSED** [review]
Code changes look good. When testing it on TS1 I was not really able to break it. Even when I added spaces directly in the database (to mimic already existing values like this) there was not really any issue. The edit exercise UI did fix this issue when saving the exercise.

File: src/test/java/de/tum/in/www1/artemis/util/PageableSearchUtilService.java [hunk 3]
```diff
@@ -115,17 +115,29 @@ public PageableSearchDTO<String> configureFinishedJobsSearchDTO() {
...
+            final Map<String, String> params = objectMapper.readValue(json, new TypeReference<>() {
+            });
+
+            // Populate a LinkedMultiValueMap from the Map, mapping parameter names to values
+            final LinkedMultiValueMap<String, String> paramMap = new LinkedMultiValueMap<>();
+            params.forEach(paramMap::add);
+            return paramMap;
+        }
+        catch (Exception e) {
+            // Handle exceptions by throwing a runtime exception with the error context
```
Comment (Hialus) [inline src/test/java/de/tum/in/www1/artemis/util/PageableSearchUtilService.java:139]: This comment is somewhat redundant.

**Review by Hialus — DISMISSED** [review]
Code still looks good. Just found two possible nitpicks.

**Review by pzdr7 — APPROVED** [review]

**Review by pzdr7 — APPROVED** [review]

**Review by Hialus — APPROVED** [review]

**Review by Strohgelaender — APPROVED** [review]

**Review by coolchock — APPROVED** [review]
tested on ts4, works as expected. Trailing and leading whitespaces were removed.

**Review by rstief — APPROVED** [review]
tested on ts4 with a variety of whitespaces. Works as expected (no whitespace is displayed after submitting and build runs through). Of course, I could not check the database on the test server.

**Review by az108 — APPROVED** [review]
Tested it on t4, with leading trails and whitespaces, also tried it in edit mode for the exercise after creating it and it still worked. And tried the steps again after deleting the exercise.
Code looks good to me

**Comment by krusche** [comment]
> Code looks good; I think the title might be in need of an update, though, as the scope is way more extensive than the PR title suggests

The title reflects the user facing changes. All others are internal and do not really affect end users. I agree that a separation into multiple PRs would have been better, but doing this now would be too much work. However, I add some details into the PR description

**Pull request merged**